pkg/core/chain/synchronizer.go

```diff
@@ -119,19 +122,31 @@ func (s *Synchronizer) ProcessBlock(m message.Message) (res []bytes.Buffer, err
 		return
 	}
 
-	if blk.Header.Height > s.highestSeen {
-		s.highestSeen = blk.Header.Height
+	s.lock.Lock()
+	currState := s.state
+	if blk.Header.Height > s.highestSeenHeight {
+		s.highestSeenHeight = blk.Header.Height
 	}
+	s.lock.Unlock()
+
+	var newState syncState
+	newState, res, err = currState(currentHeight, blk)
+
+	s.lock.Lock()
+	s.state = newState
+	s.lock.Unlock()
 
-	s.state, res, err = s.state(currentHeight, blk)
 	return
 }
 
 func (s *Synchronizer) startSync(tipHeight, currentHeight uint64) ([]bytes.Buffer, error) {
-	s.syncTarget = tipHeight
-	if s.syncTarget > currentHeight+config.MaxInvBlocks {
-		s.syncTarget = currentHeight + config.MaxInvBlocks
+
+	s.lock.Lock()
+	s.syncTargetHeight = tipHeight
+	if s.syncTargetHeight > currentHeight+config.MaxInvBlocks {
+		s.syncTargetHeight = currentHeight + config.MaxInvBlocks
 	}
+	s.lock.Unlock()
```

Comment:
This can be extrapolated into a separate function as well, since it is repeated on lines 125-130.

```golang
func (s *Synchronizer) setSyncTarget(tipHeight uint64) {
    s.lock.Lock()
    s.syncTargetHeight = tipHeight
    if s.syncTargetHeight > currentHeight+config.MaxInvBlocks {
        s.syncTargetHeight = currentHeight + config.MaxInvBlocks
    }
    s.lock.Unlock()
}
```